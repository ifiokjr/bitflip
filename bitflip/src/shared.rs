pub struct AppError(anyhow::Error);

// Tell axum how to convert `AppError` into a response.
#[cfg(feature = "ssr")]
impl axum::response::IntoResponse for AppError {
	fn into_response(self) -> axum::response::Response {
		(
			http::StatusCode::INTERNAL_SERVER_ERROR,
			format!("App error: {}", self.0),
		)
			.into_response()
	}
}

// This enables using `?` on functions that return `Result<_, anyhow::Error>` to
// turn them into `Result<_, AppError>`. That way you don't need to do that
// manually.
impl<E> From<E> for AppError
where
	E: Into<anyhow::Error>,
{
	fn from(err: E) -> Self {
		Self(err.into())
	}
}
