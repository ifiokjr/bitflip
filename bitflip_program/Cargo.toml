[package]
name = "bitflip_program"
version = "0.0.0"
edition = "2021"
description = "Created with Anchor"

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_program"

[features]
default = ["cpi"]
cpi = []
no-entrypoint = []
no-idl = []
no-log-ix-name = []
idl-build = ["anchor-lang/idl-build", "anchor-spl/idl-build"]
serde = ["dep:serde", "dep:serde_with"]
client = ["serde", "no-entrypoint", "dep:typed-builder"]

[dependencies]
anchor-lang = { workspace = true, features = ["interface-instructions", "init-if-needed"] }
anchor-spl = { workspace = true }
bitflags = { workspace = true }
bytemuck = { workspace = true }
const_str_to_pubkey = { workspace = true }
derive_more = { workspace = true, features = ["add", "add_assign"] }
serde = { workspace = true, features = ["derive"], optional = true }
serde-big-array = { workspace = true }
serde_with = { workspace = true, optional = true }
typed-builder = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
test_utils = { workspace = true }
test_utils_solana = { workspace = true }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
