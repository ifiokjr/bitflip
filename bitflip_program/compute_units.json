{
	"instructions": {
		"config_initialize": {
			"compute_units": 17569,
			"rounded_compute_units": 20000,
			"description": "Initialize the config"
		},
		"config_update_authority": {
			"compute_units": 2326,
			"rounded_compute_units": 10000,
			"description": "Update the authority of the config"
		},
		"flip_bit_test": {
			"compute_units": 64832,
			"rounded_compute_units": 70000,
			"description": "Flip a single bit from `0` to `1`"
		},
		"game_initialize": {
			"compute_units": 12979,
			"rounded_compute_units": 20000,
			"description": "Initialize the game"
		},
		"game_refresh_signer": {
			"compute_units": 2232,
			"rounded_compute_units": 10000,
			"description": "Refresh the signer"
		},
		"token_group_initialize": {
			"compute_units": 47429,
			"rounded_compute_units": 50000,
			"description": "Initialize the token group"
		},
		"token_initialize:bit": {
			"compute_units": 63184,
			"rounded_compute_units": 70000,
			"description": "Initialize group token: Bit"
		},
		"token_initialize:gibibit": {
			"compute_units": 59746,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: GibiBit"
		},
		"token_initialize:kibibit": {
			"compute_units": 59745,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: KibiBit"
		},
		"token_initialize:mebibit": {
			"compute_units": 59742,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: MebiBit"
		}
	},
	"version": "0.0.0"
}
