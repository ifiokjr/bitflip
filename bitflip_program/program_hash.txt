ebc22d77bc8ddd23f326598673e9fe7aa99fcdce9572fb8559a8d9782ade4ab8
