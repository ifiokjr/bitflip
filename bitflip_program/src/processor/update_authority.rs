use steel::*;

use super::BitflipInstruction;
use crate::BitflipError;
use crate::ConfigState;
use crate::ID;
use crate::create_pda_config;

pub fn process_update_authority(accounts: &[AccountInfo]) -> ProgramResult {
	let [config_info, authority_info, new_authority_info] = accounts else {
		return Err(ProgramError::NotEnoughAccountKeys);
	};

	// validate accounts
	config_info.is_writable()?.has_owner(&ID)?;
	authority_info
		.is_signer()?
		.is_writable()?
		.has_owner(&system_program::ID)?;
	new_authority_info
		.is_signer()?
		.has_owner(&system_program::ID)?;

	if config_info.data_len() < ConfigState::space() {
		return Err(ProgramError::AccountDataTooSmall);
	}

	let config = config_info.as_account_mut::<ConfigState>(&ID)?;
	let config_key = create_pda_config(config.bump)?;

	if config_info.key.ne(&config_key) {
		return Err(ProgramError::InvalidSeeds);
	}

	if authority_info.key.ne(&config.authority) {
		return Err(BitflipError::Unauthorized.into());
	}

	if new_authority_info.key.eq(&config.authority) {
		return Err(BitflipError::DuplicateAuthority.into());
	}

	config.authority = *new_authority_info.key;

	Ok(())
}

#[repr(C)]
#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable)]
pub struct UpdateAuthority {}

instruction!(BitflipInstruction, UpdateAuthority);

#[cfg(test)]
mod tests {
	use assert2::check;
	use solana_sdk::clock::Epoch;
	use solana_sdk::sysvar::rent::Rent;

	use super::*;
	use crate::get_pda_config;
	use crate::get_pda_mint_bit;
	use crate::get_pda_treasury;
	use crate::leak;

	#[test_log::test]
	fn should_pass() -> anyhow::Result<()> {
		let accounts = create_account_infos();
		process_update_authority(&accounts)?;

		let config_info = &accounts[0];
		let config_state = config_info.as_account::<ConfigState>(&ID)?;
		check!(config_state.authority == *accounts[2].key);

		Ok(())
	}

	#[test_log::test]
	fn should_have_enough_accounts() -> anyhow::Result<()> {
		let accounts = create_account_infos();
		let result = process_update_authority(&accounts[0..2]);

		check!(result.unwrap_err() == ProgramError::NotEnoughAccountKeys);

		Ok(())
	}

	#[test_log::test]
	fn config_must_not_be_empty() -> anyhow::Result<()> {
		let accounts = create_account_infos();
		accounts[0].realloc(0, true)?;

		let result = process_update_authority(&accounts);
		check!(result.unwrap_err() == ProgramError::AccountDataTooSmall);

		Ok(())
	}

	#[test_log::test]
	fn config_must_be_writeable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let config_info = &mut accounts[0];
		config_info.is_writable = false;

		let result = process_update_authority(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn config_must_be_pda() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let config_info = &mut accounts[0];
		config_info.key = leak(Pubkey::new_unique());

		let result = process_update_authority(&accounts);
		check!(result.unwrap_err() == ProgramError::InvalidSeeds);

		Ok(())
	}

	#[test_log::test]
	fn authority_must_be_signer() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[1];
		authority_info.is_signer = false;

		let result = process_update_authority(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn authority_must_be_writable() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[1];
		authority_info.is_writable = false;

		let result = process_update_authority(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	#[test_log::test]
	fn authority_must_be_from_config() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_info = &mut accounts[1];
		authority_info.key = leak(Pubkey::new_unique());

		let result = process_update_authority(&accounts);
		check!(result.unwrap_err() == ProgramError::Custom(BitflipError::Unauthorized.into()));

		Ok(())
	}
	#[test_log::test]
	fn new_authority_must_change() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let authority_key = *accounts[1].key;
		let new_authority_info = &mut accounts[2];
		new_authority_info.key = leak(authority_key);

		let result = process_update_authority(&accounts);
		check!(
			result.unwrap_err() == ProgramError::Custom(BitflipError::DuplicateAuthority.into())
		);

		Ok(())
	}

	#[test_log::test]
	fn new_authority_must_be_signer() -> anyhow::Result<()> {
		let mut accounts = create_account_infos();
		let new_authority_info = &mut accounts[2];
		new_authority_info.is_signer = false;

		let result = process_update_authority(&accounts);
		check!(result.unwrap_err() == ProgramError::MissingRequiredSignature);

		Ok(())
	}

	fn create_account_infos() -> [AccountInfo<'static>; 3] {
		let (config_key, config_bump) = leak(get_pda_config());
		let treasury_bump = get_pda_treasury().1;
		let mint_bump = get_pda_mint_bit().1;
		let authority_lamports = leak(1_000_000_000);
		let new_authority_lamports = leak(1_000_000_000);
		let authority_key = leak(Pubkey::new_unique());
		let new_authority_key = leak(Pubkey::new_unique());
		let rent_sysvar = Rent::default();
		let minimum_balance = leak(rent_sysvar.minimum_balance(ConfigState::space()));
		let mut data = vec![0u8; 8];
		data[0] = ConfigState::discriminator();
		data.append(
			&mut ConfigState::new(*authority_key, *config_bump, treasury_bump, mint_bump)
				.to_bytes()
				.to_vec(),
		);
		let authority_data = leak(vec![]);
		let new_authority_data = leak(vec![]);

		let config_info = AccountInfo::new(
			config_key,
			false,
			true,
			minimum_balance,
			leak(data),
			&ID,
			false,
			Epoch::default(),
		);
		let authority_info = AccountInfo::new(
			authority_key,
			true,
			true,
			authority_lamports,
			authority_data,
			&system_program::ID,
			false,
			Epoch::default(),
		);
		let new_authority_info = AccountInfo::new(
			new_authority_key,
			true,
			false,
			new_authority_lamports,
			new_authority_data,
			&system_program::ID,
			false,
			Epoch::default(),
		);

		[config_info, authority_info, new_authority_info]
	}
}
