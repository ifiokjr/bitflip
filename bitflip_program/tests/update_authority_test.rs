#![cfg(feature = "client")]

use std::future::Future;

use assert2::check;
use bitflip_program::BitflipError;
use bitflip_program::ConfigState;
use bitflip_program::get_pda_config;
use bitflip_program::update_authority;
use shared::ToRpcClient;
use shared::create_authority_keypair;
use shared::create_config_state;
use shared::create_program_context_with_factory;
use solana_sdk::instruction::InstructionError;
use solana_sdk::signature::Keypair;
use solana_sdk::transaction::TransactionError;
use solana_sdk::transaction::VersionedTransaction;
use steel::*;
use test_utils::create_insta_redaction;
use test_utils_solana::prelude::*;

mod shared;

#[test_log::test(tokio::test)]
async fn update_authority_test() -> anyhow::Result<()> {
	shared_update_authority_test(create_banks_client_rpc).await?;
	Ok(())
}

#[cfg(feature = "test_validator")]
#[test_log::test(tokio::test)]
async fn update_authority_test_validator() -> anyhow::Result<()> {
	let compute_units = shared_update_authority_test(create_validator_rpc).await?;
	let rounded_compute_units = bitflip_program::round_compute_units_up(compute_units);

	check!(rounded_compute_units == 10_000);
	insta::assert_snapshot!(format!("{rounded_compute_units} CU"));

	Ok(())
}

#[test_log::test(tokio::test)]
async fn authority_must_change_test() -> anyhow::Result<()> {
	shared_authority_must_change_test(create_banks_client_rpc).await
}

#[cfg(feature = "test_validator")]
#[test_log::test(tokio::test)]
async fn authority_must_change_test_validator() -> anyhow::Result<()> {
	shared_authority_must_change_test(create_validator_rpc).await
}

async fn create_banks_client_rpc() -> anyhow::Result<impl ToRpcClient> {
	let provider = create_program_context_with_factory(|p| {
		let config = get_pda_config().0;
		let config_state_account = create_config_state();
		p.add_account(config, config_state_account.into());
	})
	.await?;

	Ok(provider)
}

#[cfg(feature = "test_validator")]
async fn create_validator_rpc() -> anyhow::Result<impl ToRpcClient> {
	let mut accounts = std::collections::HashMap::new();
	let config = get_pda_config().0;
	let config_state_account = create_config_state();
	accounts.insert(config, config_state_account);

	let runner = shared::create_runner_with_accounts(accounts).await;

	Ok(runner)
}

async fn shared_update_authority_test<
	T: ToRpcClient,
	Fut: Future<Output = anyhow::Result<T>>,
	Create: FnOnce() -> Fut,
>(
	create_provider: Create,
) -> anyhow::Result<u64> {
	let provider = create_provider().await?;
	let rpc = provider.to_rpc();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let new_authority_keypair = Keypair::new();
	let new_authority = new_authority_keypair.pubkey();
	let config = get_pda_config().0;
	let recent_blockhash = rpc.get_latest_blockhash().await?;
	let ix = update_authority(&authority, &new_authority);
	let mut transaction =
		VersionedTransaction::new_unsigned_v0(&authority, &[ix], &[], recent_blockhash)?;

	let simulation = rpc.simulate_transaction(&transaction).await?;
	log::info!("simulation: {simulation:#?}");
	let compute_units = simulation.value.units_consumed.unwrap();

	transaction
		.try_sign(&[&new_authority_keypair], None)?
		.try_sign(&[&authority_keypair], None)?;

	let signature = rpc.send_and_confirm_transaction(&transaction).await?;
	rpc.confirm_transaction(&signature).await?;

	let config_state_data = rpc.get_account_data(&config).await?;
	let config_state_account = ConfigState::try_from_bytes(&config_state_data)?;
	let authority_redaction = create_insta_redaction(new_authority, "new_authority:pubkey");
	insta::assert_compact_json_snapshot!(config_state_account,{
		".authority" => insta::dynamic_redaction(authority_redaction),
	}, @r#"
 {
   "authority": "[new_authority:pubkey]",
   "bump": 254,
   "treasuryBump": 255,
   "mintBitBump": 255,
   "mintKibibitBump": 255,
   "mintMebibitBump": 255,
   "mintGibibitBump": 255,
   "gameIndex": 0
 }
 "#);

	Ok(compute_units)
}

async fn shared_authority_must_change_test<
	T: ToRpcClient,
	Fut: Future<Output = anyhow::Result<T>>,
	Create: FnOnce() -> Fut,
>(
	create_provider: Create,
) -> anyhow::Result<()> {
	let provider = create_provider().await?;
	let rpc = provider.to_rpc();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let recent_blockhash = rpc.get_latest_blockhash().await?;
	let ix = update_authority(&authority, &authority);
	let transaction =
		VersionedTransaction::new_unsigned_v0(&authority, &[ix], &[], recent_blockhash)?;
	let simulation = rpc.simulate_transaction(&transaction).await?;

	log::info!("simulation: {simulation:#?}");
	check!(
		simulation.value.err.unwrap()
			== TransactionError::InstructionError(
				0,
				InstructionError::Custom(BitflipError::DuplicateAuthority.into())
			)
	);

	Ok(())
}
