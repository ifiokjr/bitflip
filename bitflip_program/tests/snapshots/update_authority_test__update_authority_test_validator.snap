---
source: bitflip_program/tests/update_authority_test.rs
expression: "format!(\"{rounded_compute_units} CU\")"
snapshot_kind: text
---
10000 CU
